{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 5: Link Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this assignment, we will use network-based algorithms, such as PageRank, to improve our search results. After the last assignment on evaluating IR systems, we go back to our PubMed dataset of scientific papers. In this dataset, we look at two graphs in particular: the co-authorship network and the citation network.\n",
    "\n",
    "The citation network is similar to the link network of the web: Citations are like web links pointing to other documents. We can therefore apply the same network-based ranking methods."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code from previous assignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle, bz2\n",
    "from collections import defaultdict, namedtuple, Counter\n",
    "from math import log10, sqrt\n",
    "from IPython.display import display, HTML\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# show plots inline within the notebook\n",
    "%matplotlib inline\n",
    "# set plots' resolution\n",
    "plt.rcParams['savefig.dpi'] = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ids_file = 'data/fungus_Ids.pkl.bz2'\n",
    "Summaries_file = 'data/fungus_Summaries.pkl.bz2'\n",
    "Citations_file = 'data/fungus_Citations.pkl.bz2'\n",
    "Abstracts_file = 'data/fungus_Abstracts.pkl.bz2'\n",
    "\n",
    "Ids = pickle.load( bz2.BZ2File( Ids_file, 'rb' ) )\n",
    "Summaries = pickle.load( bz2.BZ2File( Summaries_file, 'rb' ) )\n",
    "Citations = pickle.load( bz2.BZ2File( Citations_file, 'rb' ) )\n",
    "Abstracts = pickle.load( bz2.BZ2File( Abstracts_file, 'rb' ) )\n",
    "\n",
    "paper = namedtuple( 'paper', ['title', 'authors', 'year', 'doi'] )\n",
    "\n",
    "for (id, paper_info) in Summaries.items():\n",
    "    Summaries[id] = paper( *paper_info )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_summary( id, show_abstract=False, show_id=True, extra_text='' ):\n",
    "    \"\"\"\n",
    "    Function for printing a paper's summary through IPython's Rich Display System.\n",
    "    Trims long author lists, and adds a link to the paper's DOI (when available).\n",
    "    \"\"\"\n",
    "    s = Summaries[id]\n",
    "    lines = []\n",
    "    title = s.title\n",
    "    if s.doi != '':\n",
    "        title = '<a href=http://dx.doi.org/{:s}>{:s}</a>'.format(s.doi, title)\n",
    "    title = '<strong>' + title + '</strong>'\n",
    "    lines.append(title)\n",
    "    authors = ', '.join( s.authors[:20] ) + ('' if len(s.authors) <= 20 else ', ...')\n",
    "    lines.append(str(s.year) + '. ' + authors)\n",
    "    if (show_abstract):\n",
    "        lines.append('<small><strong>Abstract:</strong> <em>{:s}</em></small>'.format(Abstracts[id]))\n",
    "    if (show_id):\n",
    "        lines.append('[ID: {:d}]'.format(id))\n",
    "    if (extra_text != ''):\n",
    "         lines.append(extra_text)\n",
    "    display( HTML('<br>'.join(lines)) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tokenize(text):\n",
    "    return text.split(' ')\n",
    "\n",
    "def preprocess(tokens):\n",
    "    result = []\n",
    "    for token in tokens:\n",
    "        result.append(token.lower())\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "inverted_index = defaultdict(list)\n",
    "\n",
    "for id in sorted(Summaries.keys()):\n",
    "    term_set = set(preprocess(tokenize(Summaries[id].title)))\n",
    "    if id in Abstracts:\n",
    "        term_set.update(preprocess(tokenize(Abstracts[id])))\n",
    "    for term in term_set:\n",
    "        inverted_index[term].append(id)\n",
    "\n",
    "tf_matrix = defaultdict(Counter)\n",
    "\n",
    "for doc_id in Summaries.keys():\n",
    "    tokens = preprocess(tokenize(Summaries[doc_id].title))\n",
    "    if (doc_id in Abstracts):\n",
    "        tokens.extend(preprocess(tokenize(Abstracts[doc_id])))\n",
    "    tf_matrix[doc_id] = Counter(tokens)\n",
    "\n",
    "def tf(t,d):\n",
    "    return float(tf_matrix[d][t])\n",
    "\n",
    "def df(t):\n",
    "    return float(len(inverted_index[t]))\n",
    "\n",
    "num_documents = float(len(Summaries))\n",
    "\n",
    "def idf(t):\n",
    "    return log10((num_documents + 1)/(df(t) + 1))\n",
    "\n",
    "def tfidf(t,d):\n",
    "    return tf(t,d) * idf(t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Co-authorship network"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start by building a mapping from authors to the set of identifiers of papers they authored.  We'll be using Python sets again for that purpose."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "papers_of_author = defaultdict(set)\n",
    "\n",
    "for (id, p) in Summaries.items():\n",
    "    for a in p.authors:\n",
    "        papers_of_author[a].add(id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's try it out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{25154107, 28425998, 29491409, 30868671}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "papers_of_author['Canessa S']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/s41598-018-22225-9>Post-epizootic salamander persistence in a disease-free refugium suggests poor dispersal ability of Batrachochytrium salamandrivorans.</a></strong><br>2018. Spitzen-van der Sluijs A, Stegen G, Bogaerts S, Canessa S, Steinfartz S, Janssen N, Bosman W, Pasmans F, Martel A<br>[ID: 29491409]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1890/13-0103.1>Designing screening protocols for amphibian disease that account for imperfect and variable capture rates of individuals.</a></strong><br>2014. Canessa S, Martel A, Pasmans F<br>[ID: 25154107]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nature22059>Drivers of salamander extirpation mediated by Batrachochytrium salamandrivorans.</a></strong><br>2017. Stegen G, Pasmans F, Schmidt BR, Rouffaer LO, Van Praet S, Schaub M, Canessa S, Laudelout A, Kinet T, Adriaensen C, Haesebrouck F, Bert W, Bossuyt F, Martel A<br>[ID: 28425998]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1111/cobi.13313>Quantifying the burden of managing wildlife diseases in multiple host species.</a></strong><br>2019. Canessa S, Bozzuto C, Pasmans F, Martel A<br>[ID: 30868671]"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for id in papers_of_author['Canessa S']:\n",
    "    display_summary(id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now build a co-authorship network, that is a graph linking authors to the set of co-authors they have published with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "coauthors = defaultdict(set)\n",
    "\n",
    "for p in Summaries.values():\n",
    "    for a in p.authors:\n",
    "        coauthors[a].update(p.authors)\n",
    "\n",
    "# The code above results in each author being listed as having co-authored with himself/herself.\n",
    "# We remove these self-references here:\n",
    "for (a, ca) in coauthors.items():\n",
    "    ca.remove(a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And let's try it out again:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Spitzen-van der Sluijs A, Janssen N, Bossuyt F, Martel A, Schmidt BR, Bosman W, Bozzuto C, Bert W, Stegen G, Laudelout A, Rouffaer LO, Bogaerts S, Schaub M, Steinfartz S, Van Praet S, Adriaensen C, Haesebrouck F, Kinet T, Pasmans F\n"
     ]
    }
   ],
   "source": [
    "print(', '.join( coauthors['Canessa S'] ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unlike a citation or link network, the edges of this co-authorship network are not directed: There is no direction (no arrow) in the link between author 'Canessa S' and 'Schmidt BR', for example. With our chosen implementation, each of these links in fact appears twice in our data, as we also get 'Canessa S' as co-author when we look for 'Schmidt BR':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bales EK, Goverse E, Cashins S, Puschendorf R, Vörös J, Zamudio KR, Muths E, Skerratt LF, Farrer RA, Kolby JE, Bataille A, Nishikawa K, James TY, Aanensen DM, Van Bocxlaer I, Lauer A, Doddington BJ, Ghosh P, Laudelout A, Wierzbicki C, Tobler U, Wombwell E, Oliver JA, Courtois EA, Ohlhoff D, Steinfartz S, Minting P, Bates KA, Pasmans F, Toledo LF, Voyles J, Miaud C, Spikmans F, Spitzen-van der Sluijs A, Bossuyt F, Martel A, Blooi M, Loyau A, O'Hanlon SJ, Clare F, Verster R, Rosa GM, Martin MD, Speybroeck J, Nöllert A, Canessa S, Waldman B, Bozzuto C, Garner TW, Bert W, Woodhams DC, Höglund J, Nguyen TT, Shelton JMG, Spagnoletti M, Rouffaer LO, Wagner N, Schaub M, Beukema W, Brankovics B, Garcia G, Geiger CC, Marin da Fonte LF, Kerres A, Murray KA, Bosch J, Kinet T, Fisher MC, Ducatelle R, Smith F, Weldon C, Doherty-Bone TM, Sheafor B, Sabino-Pinto J, Muletz C, Schmeller DS, Jenkinson TS, Asselberghs J, Stegen G, Lötters S, Dalbeck L, Gilbert MTP, Van Praet S, Fumagalli M, Bletz MC, Adriaensen C, Salvidio S, Hintz WE, Valenzuela-Sánchez A, Soto-Azat C, Laurila A, Van Rooij P, Böll S, Veith M, Borgula A, Alvarado-Rybak M, Goka K, Lips KR, Grassly NC, Balloux F, Rieux A, Webb RJ, Wales N, Briggs CJ, Tessa G, Brookes L, Berger L, Kosch TA, Lin CF, Vences M, Meurling S, Gower DJ, Cunningham AA, Haesebrouck F, Davis LR, Kirst K, Garner TWJ\n"
     ]
    }
   ],
   "source": [
    "print(', '.join( coauthors['Schmidt BR'] ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With that in mind, we can calculate some basic statistics about our graph:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes (authors):  120759\n",
      "Number of links (co-authorship relations):  708774\n"
     ]
    }
   ],
   "source": [
    "print('Number of nodes (authors): ', len(coauthors))\n",
    "\n",
    "# We divide by two here to account for the fact that each edge is represented twice (see above):\n",
    "coauthor_rel_count = int(sum( len(c) for c in coauthors.values() ) / 2)\n",
    "print('Number of links (co-authorship relations): ', coauthor_rel_count)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this data at hand, we can plot the [degree distribution](https://en.wikipedia.org/wiki/Degree_distribution) by showing the number of collaborators a scientist has published with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAEKCAYAAAA4t9PUAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAeKklEQVR4nO3de7QdZZ3m8e/TQQOiXAIB0wkxQdK2QCPKGYwtbSOoxIYhuFo0jkpUxoxIA0rbCroU2lnphasdUWYGNAISFMF4JV5QMDRqt1xMEAkXkbSJEIkEFDEDGkl45o96j2xO9jmpcyp7H3b281lrr131q9tvF6zzS71v1VuyTURExFj92XgnEBERvS2FJCIiGkkhiYiIRlJIIiKikRSSiIhoJIUkIiIa6VghkXSxpPWSbmuz7D2SLGnPltiZklZJukvSUS3xQyStLMvOk6QSnyjpCyV+o6QZnfotERExvE5ekVwCzBkalLQP8ErgnpbY/sA84ICyzfmSJpTFFwALgFnlM7jPE4GHbO8HnAt8pCO/IiIiRtSxQmL7+8Bv2iw6F3gv0Pok5FzgCtsbba8GVgGHSpoC7GL7eldPTl4KHNeyzeIy/SXgyMGrlYiI6J4dunkwSccCv7T9kyF/86cCN7TMry2xx8r00PjgNvcC2N4k6WFgD+DBkXLYc889PWPGjAa/IiKi/6xYseJB25PbLetaIZH0DOADwKvaLW4T8wjxkbZpd+wFVM1jTJ8+neXLl28134iIeIKkXwy3rJt3bT0XmAn8RNIaYBpws6RnU11p7NOy7jTgvhKf1iZO6zaSdgB2pX1TGrYX2R6wPTB5ctuCGhERY9S1QmJ7pe29bM+wPYOqELzI9q+ApcC8cifWTKpO9ZtsrwM2SJpd+j9OAK4su1wKzC/TrwWudUagjIjouk7e/ns5cD3wPElrJZ043Lq2bweWAHcA3wZOtr25LD4JuJCqA/4/gatK/CJgD0mrgNOBMzryQyIiYkTqt3/EDwwMOH0kERGjI2mF7YF2y/Jke0RENJJCEhERjaSQREREIykkERHRSApJREQ00tUhUvrJjDO+OeyyNecc3cVMIiI6K1ckERHRSApJREQ0kkISERGNpJBEREQjKSQREdFICklERDSSQhIREY2kkERERCMpJBER0UgKSURENJIhUsbBcMOnZOiUiOhFuSKJiIhGUkgiIqKRFJKIiGgkhSQiIhpJIYmIiEY6VkgkXSxpvaTbWmL/Kumnkm6V9FVJu7UsO1PSKkl3STqqJX6IpJVl2XmSVOITJX2hxG+UNKNTvyUiIobXySuSS4A5Q2LXAAfaPgj4GXAmgKT9gXnAAWWb8yVNKNtcACwAZpXP4D5PBB6yvR9wLvCRjv2SiIgYVscKie3vA78ZErva9qYyewMwrUzPBa6wvdH2amAVcKikKcAutq+3beBS4LiWbRaX6S8BRw5erURERPeMZx/J24CryvRU4N6WZWtLbGqZHhp/0jalOD0M7NHBfCMioo1xebJd0geATcBlg6E2q3mE+EjbtDveAqrmMaZPnz6qXLdmuKfUIyL6RdevSCTNB44B3liaq6C60tinZbVpwH0lPq1N/EnbSNoB2JUhTWmDbC+yPWB7YPLkydvqp0REBF0uJJLmAO8DjrX9aMuipcC8cifWTKpO9ZtsrwM2SJpd+j9OAK5s2WZ+mX4tcG1LYYqIiC7pWNOWpMuBw4E9Ja0FzqK6S2sicE3pF7/B9jts3y5pCXAHVZPXybY3l12dRHUH2E5UfSqD/SoXAZ+VtIrqSmRep35LREQMr2OFxPYb2oQvGmH9hcDCNvHlwIFt4n8Ajm+SY0RENJcn2yMiopEUkoiIaCSFJCIiGkkhiYiIRlJIIiKikRSSiIhoJIUkIiIaSSGJiIhGUkgiIqKRFJKIiGgkhSQiIhpJIYmIiEZSSCIiopEUkoiIaCSFJCIiGhmXd7ZHeyO9/33NOUd3MZOIiPq2ekUi6bmSJpbpwyWdKmm3zqcWERG9oE7T1peBzZL2o3rD4Uzg8x3NKiIiekadQvK47U3Aa4CP2343MKWzaUVERK+oU0gek/QGYD7wjRJ7WudSioiIXlKnkLwVeAmw0PZqSTOBz3U2rYiI6BUj3rUlaQLwfttvGozZXg2c0+nEIiKiN4x4RWJ7MzBZ0tNHu2NJF0taL+m2ltgkSddIurt8796y7ExJqyTdJemolvghklaWZedJUolPlPSFEr9R0ozR5hgREc3VadpaA/yHpA9KOn3wU2O7S4A5Q2JnAMtszwKWlXkk7Q/MAw4o25xfroYALgAWALPKZ3CfJwIP2d4POBf4SI2cIiJiG6tTSO6j6mT/M+BZLZ8R2f4+8Jsh4bnA4jK9GDiuJX6F7Y2l6WwVcKikKcAutq+3beDSIdsM7utLwJGDVysREdE9W32y3fY/A0ja2fYjDY+3t+11Zb/rJO1V4lOBG1rWW1tij5XpofHBbe4t+9ok6WFgD+DBhjlGRMQo1Hmy/SWS7gDuLPMvkHT+Ns6j3ZWER4iPtM2WO5cWSFouafkDDzwwxhQjIqKdOk1bHweOAn4NYPsnwMvGeLz7S3MV5Xt9ia8F9mlZbxpVk9raMj00/qRtJO0A7MqWTWmUnBfZHrA9MHny5DGmHhER7dQa/df2vUNCm8d4vKVUDzZSvq9sic8rd2LNpOpUv6k0g22QNLv0f5wwZJvBfb0WuLb0o0RERBfVGf33Xkl/DbjcBnwqpZlrJJIuBw4H9pS0FjiL6vmTJZJOBO4BjgewfbukJcAdwCbg5HLrMcBJVHeA7QRcVT5Qjfv1WUmrqK5E5tX4LRERsY3VKSTvAD5B1bm9FrgaOHlrG9l+wzCLjhxm/YXAwjbx5cCBbeJ/oBSiiIgYP3Xu2noQeGMXcomIiB601UIiaTLwdmBG6/q239a5tCIiolfUadq6EvgB8F3G3skeERHbqTqF5Bm239fxTCIioifVuf33G5L+ruOZRERETxr2ikTSBp54uvz9kjZSDVkiwLZ36U6KERHxVDZsIbG91YEZIyIi6oy19RpJu7bM7ybpuJG2iYiI/lGnj+Qs2w8Pztj+LdVT6hEREbUKSbt16tztFRERfaBOIVku6WOSnitpX0nnAis6nVhERPSGOoXkFOCPwBeAJcDvqTHWVkRE9IcRm6jKe9PPtv1PXconIiJ6zIiFxPZmSYd0K5mnqhlnfHO8U4iIeMqq02n+Y0lLgS8Cf3pnu+2vdCyriIjoGXUKySSq1+we0RIzkELSRSNdFa055+guZhIR8WR13kfy1m4kEhERvanO+0h2BE4EDgB2HIznfSQREQH1bv/9LPBs4Cjge8A0YEMnk4qIiN5Rp5DsZ/uDwCO2FwNHA3/V2bQiIqJX1Ckkj5Xv30o6ENiV6rW7ERERte7aWiRpd+CDwFLgmcCHOppVRET0jK1ekdi+0PZDtr9ne1/be9n+ZJODSnq3pNsl3Sbpckk7Spok6RpJd5fv3VvWP1PSKkl3STqqJX6IpJVl2XmS1CSviIgYvTrvI9lb0kWSrirz+0s6cawHlDQVOBUYsH0gMAGYB5wBLLM9C1hW5pG0f1l+ADAHOL8M3QJwAbAAmFU+c8aaV0REjE2dPpJLgO8Af17mfwa8q+FxdwB2krQD8AzgPmAusLgsXwwMvjxrLnCF7Y22VwOrgEMlTQF2sX29bQOXtmwTERFdUqeQ7Gl7CfA4gO1NwOaxHtD2L4GPAvcA64CHbV8N7G17XVlnHbBX2WQqcG/LLtaW2NQyPTQeERFdVKeQPCJpD6phUZA0G3h45E2GV/o+5gIzqa5ydpb0ppE2aRPzCPF2x1wgabmk5Q888MBoU46IiBHUKSSnU92t9VxJ/0HVhHRKg2O+Alht+wHbj1GN2fXXwP2luYryvb6svxbYp2X7aVRNYWvL9ND4Fmwvsj1ge2Dy5MkNUo+IiKHq3LV1M/C3VH/s/wdwgO1bGxzzHmC2pGeUu6yOBO6kKlbzyzrzgSvL9FJgnqSJkmZSdarfVJq/NkiaXfZzQss2ERHRJXXu2joe2Mn27VSd2V+Q9KKxHtD2jcCXgJuBlSWHRcA5wCsl3Q28ssxTjrsEuAP4NnCy7cE+mpOAC6k64P8TuGqseUVExNjUeSDxg7a/KOkwqvG2Pkp12+2Lx3pQ22cBZw0Jb6S6Omm3/kJgYZv4cuDAseYRERHN1ekjGfzX/9HABbavBJ7euZQiIqKX1Ckkv5T0KeB1wLckTay5XURE9IE6BeF1VA8kzrH9W6o3Jv5TR7OKiIieUeeurUepbsU9rIQ2AXd3MqmIiOgdde7aOgt4H3BmCT0N+Fwnk4qIiN5Rp2nrNcCxwCMAtu8DntXJpCIionfUKSR/LIMiDg6RsnNnU4qIiF5Sp5AsKXdt7Sbp7cB3gU93Nq2IiOgVW30g0fZHJb0S+B3wPOBDtq/peGYREdETRiwk5QVS37H9CiDFIyIitjBi01YZ0+pRSbt2KZ+IiOgxdcba+gOwUtI1lDu3AGyf2rGsIiKiZ9QpJN8sn4iIiC3U6WxfvLV1YnzNOKN9nV9zztFdziQi+lEGX4yIiEZSSCIiopFhC4mkz5bv07qXTkRE9JqRrkgOkfQc4G2Sdpc0qfXTrQQjIuKpbaTO9k9SvSN9X2AFoJZlLvGIiOhzw16R2D7P9vOBi23va3tmyydFJCIigHq3/54k6QXA35TQ923f2tm0IiKiV9R5sdWpwGXAXuVzmaRTOp1YRET0hjq3//534MW2P2T7Q8Bs4O1NDippN0lfkvRTSXdKeknpxL9G0t3le/eW9c+UtErSXZKOaokfImllWXaeJLU/YkREdEqdQiJgc8v8Zp7c8T4WnwC+bfsvgRcAdwJnAMtszwKWlXkk7Q/MAw4A5gDnl1GJAS4AFgCzymdOw7wiImKU6oy19RngRklfLfPHAReN9YCSdgFeBrwFwPYfgT9KmgscXlZbDFxH9a74ucAVtjcCqyWtAg6VtAbYxfb1Zb+XltyuGmtuERExenU62z8m6TrgMKorkbfa/nGDY+4LPAB8pnTirwBOA/a2va4cc52kvcr6U4EbWrZfW2KPlemh8YiI6KI6VyTYvhm4eRse80XAKbZvlPQJSjPWMNo1o3mE+JY7kBZQNYExffr00WUbEREjqlVItrG1wFrbN5b5L1EVkvslTSlXI1OA9S3r79Oy/TTgvhKf1ia+BduLgEUAAwMDbYvN9mi4UYEhIwNHxLbT9UEbbf8KuFfS80roSOAOYCkwv8TmA1eW6aXAPEkTJc2k6lS/qTSDbZA0u9ytdULLNhER0SWjeWf7tnQK1fMoTwd+DryVqqgtkXQicA9wPIDt2yUtoSo2m4CTyyuAAU4CLgF2oupkT0d7RESXjVhIbG+W9KikXW0/vK0OavsWYKDNoiOHWX8hsLBNfDlw4LbKKyIiRi/vbI+IiEbyzvaIiGik1jvbJe0ETLd9VxdyioiIHlJn0Mb/CtxC9W4SJB0saWmnE4uIiN5Q5/bfs4FDgd/CnzrKZ3Ywp4iI6CF1CsmmNnds9c1DfRERMbI6ne23SfpvwARJs4BTgR92Nq2IiOgVda5ITqEawn0jcDnwO+BdnUwqIiJ6R527th4FPiDpI9WsN3Q+rYiI6BV17tr6L5JWArdSPZj4E0mHdD61iIjoBXX6SC4C3mn7BwCSDqN62dVBnUxsPIw0Wm5ERLRXp49kw2ARAbD970CatyIiAhjhikTSi8rkTZI+RdXRbuD1VK/BjYiIGLFp638NmT+rZTrPkUREBDBCIbH98m4mEhERvWmrne2SdqN6++CM1vUzjHxEREC9u7a+BdwArAQe72w6ERHRa+oUkh1tn97xTCIioifVuf33s5LeLmmKpEmDn45nFhERPaHOFckfgX8FPsATd2sZ2LdTSUXnDffw5Zpzju5yJhHR6+oUktOB/Ww/2OlkIiKi99Rp2rodeLTTiURERG+qU0g2A7dI+pSk8wY/TQ8saYKkH0v6RpmfJOkaSXeX791b1j1T0ipJd0k6qiV+iKSVZdl5ktQ0r4iIGJ06heRrwEKql1mtaPk0dRpwZ8v8GcAy27OAZWUeSfsD86jeiTIHOF/ShLLNBcACYFb5zNkGeUVExCjUeR/J4m19UEnTgKOpCtTgrcVzgcPL9GKq8bzeV+JX2N4IrJa0CjhU0hpgF9vXl31eChwHXLWt842IiOHVebJ9NW3G1rLd5K6tjwPvBZ7VEtvb9rqy73WS9irxqVQPRA5aW2KPlemh8YiI6KI6d20NtEzvCBwPjPk5EknHAOttr5B0eJ1N2sQ8QrzdMRdQNYExffr0mplGREQdW+0jsf3rls8vbX8cOKLBMV8KHFuapq4AjpD0OeB+SVMAyvf6sv5aYJ+W7acB95X4tDbxdr9hke0B2wOTJ09ukHpERAxV51W7L2r5DEh6B09ukhoV22fanmZ7BlUn+rW23wQsBeaX1eYDV5bppcA8SRMlzaTqVL+pNINtkDS73K11Qss2ERHRJXWatlrfS7IJWAO8rgO5nAMskXQicA9VExq2b5e0BLijHP9k25vLNicBlwA7UXWyp6M9IqLL6ty11bH3kti+jvK2Rdu/Bo4cZr2FVHd4DY0vBw7sVH4REbF1de7amgj8PVu+j+TDnUsrIiJ6RZ2mrSuBh6keQtzY2XQiIqLX1Ckk02znifGIiGirTiH5oaS/sr2y49nEuBtueHnIEPMR0V6dQnIY8JbyhPtGqgcBbfugjmYWERE9oU4heXXHs4iIiJ5V5/bfX3QjkYiI6E11hpGPiIgYVgpJREQ0kkISERGNpJBEREQjde7aigCGf8Ykz5dE9LdckURERCMpJBER0UgKSURENJJCEhERjaSQREREIykkERHRSApJREQ0kkISERGNpJBEREQjKSQREdFI14dIkbQPcCnwbOBxYJHtT0iaBHwBmAGsAV5n+6GyzZnAicBm4FTb3ynxQ4BLgJ2AbwGn2XY3f0/k9bwR/W48rkg2Af9o+/nAbOBkSfsDZwDLbM8ClpV5yrJ5wAHAHOB8SRPKvi4AFgCzymdON39IRESMQyGxvc72zWV6A3AnMBWYCywuqy0GjivTc4ErbG+0vRpYBRwqaQqwi+3ry1XIpS3bREREl4xrH4mkGcALgRuBvW2vg6rYAHuV1aYC97ZstrbEppbpofGIiOiicRtGXtIzgS8D77L9O0nDrtom5hHi7Y61gKoJjOnTp48+2Riz9J9EbP/G5YpE0tOoishltr9SwveX5irK9/oSXwvs07L5NOC+Ep/WJr4F24tsD9gemDx58rb7IRER0f1CourS4yLgTtsfa1m0FJhfpucDV7bE50maKGkmVaf6TaX5a4Ok2WWfJ7RsExERXTIeTVsvBd4MrJR0S4m9HzgHWCLpROAe4HgA27dLWgLcQXXH18m2N5ftTuKJ23+vKp+IiOiirhcS2/9O+/4NgCOH2WYhsLBNfDlw4LbLLiIiRivvbI9xk3fAR2wf+q6QrPzlwyPeSRQREaPTd4Uknvpyy3BEb8mgjRER0UgKSURENJJCEhERjaSPJHpK7vSKeOrJFUlERDSSQhIREY2kkERERCPpI4ntQp49iRg/KSSx3UsHfURnpWkrIiIayRVJ9K00h0VsG7kiiYiIRnJFEtFG+lUi6kshiRiFbf0KghSm2B6kaSsiIhrJFUnEOBrLFU6uYuKpJlckERHRSK5IInpMN18VnaufqCOFJCJGLc/gRCvZHu8cGpE0B/gEMAG40PY5I60/ccosT5n/8a7kFhH1bOvik9u3tz1JK2wPtF3Wy4VE0gTgZ8ArgbXAj4A32L5juG1SSCKi07pVGDtxrOGMVEh6vWnrUGCV7Z8DSLoCmAsMW0giIjqtm/1Y2/JYYy1KvV5IpgL3tsyvBV48TrlERPS0sRalXi8kahPboq1O0gJgQZnd+IuPHHNbR7PqDXsCD453EuMs56CS81DJeRj5HDxnuI16vZCsBfZpmZ8G3Dd0JduLgEUAkpYP187XT3Iecg4G5TxUch7Gfg56/YHEHwGzJM2U9HRgHrB0nHOKiOgrPX1FYnuTpH8AvkN1++/Ftm8f57QiIvpKTxcSANvfAr41ik0WdSqXHpPzkHMwKOehkvMwxnPQ08+RRETE+Ov1PpKIiBhnfVVIJM2RdJekVZLOGO98ukHSxZLWS7qtJTZJ0jWS7i7fu49njt0gaR9J/ybpTkm3SzqtxPvmXEjaUdJNkn5SzsE/l3jfnINWkiZI+rGkb5T5vjsPktZIWinpFknLS2zU56FvCkkZTuX/Aq8G9gfeIGn/8c2qKy4B5gyJnQEssz0LWFbmt3ebgH+0/XxgNnBy+e/fT+diI3CE7RcABwNzJM2mv85Bq9OAO1vm+/U8vNz2wS23/Y76PPRNIaFlOBXbfwQGh1PZrtn+PvCbIeG5wOIyvRg4rqtJjQPb62zfXKY3UP0BmUofnQtX/l+ZfVr5mD46B4MkTQOOBi5sCffdeRjGqM9DPxWSdsOpTB2nXMbb3rbXQfUHFthrnPPpKkkzgBcCN9Jn56I059wCrAeusd1356D4OPBe4PGWWD+eBwNXS1pRRgCBMZyHnr/9dxRqDacS2zdJzwS+DLzL9u+kdv9bbL9sbwYOlrQb8FVJB453Tt0m6Rhgve0Vkg4f73zG2Utt3ydpL+AaST8dy0766Yqk1nAqfeJ+SVMAyvf6cc6nKyQ9jaqIXGb7KyXcl+fC9m+B66j6z/rtHLwUOFbSGqom7iMkfY7+Ow/Yvq98rwe+StUFMOrz0E+FJMOpPGEpML9MzweuHMdcukLVpcdFwJ22P9ayqG/OhaTJ5UoESTsBrwB+Sh+dAwDbZ9qeZnsG1d+Ba22/iT47D5J2lvSswWngVcBtjOE89NUDiZL+jqptdHA4lYXjnFLHSbocOJxqVM/7gbOArwFLgOnAPcDxtod2yG9XJB0G/ABYyRPt4u+n6ifpi3Mh6SCqztMJVP+IXGL7w5L2oE/OwVClaes9to/pt/MgaV+qqxCoujk+b3vhWM5DXxWSiIjY9vqpaSsiIjoghSQiIhpJIYmIiEZSSCIiopEUkoiIaCSFJKKFpOskdfy93ZJOLSMRX9bpY20lj7dI+vOW+TWS9hzPnKL39NMQKREdJWkH25tqrv5O4NW2V3cypxreQvUQWuNRHkb5+2M7kiuS6DmSZpR/zX+6vFfj6vKk9pOuKCTtWYbBGPyX99ckfV3Sakn/IOn08j6KGyRNajnEmyT9UNJtkg4t2++s6t0uPyrbzG3Z7xclfR24uk2up5f93CbpXSX2SWBfYKmkdw9Zf4Kkj5Z3RNwq6ZQSP7Icd2XJY2KbYz1T0jJJN5f1BnOcoSe/j+Y9ks6W9FpgALisvI9ip7LKKS37+MuyzaRy/m4t5+ugEj9b0iJJVwOXSjpA1TtPbinrzhrVf9zoTbbzyaenPsAMqveLHFzmlwBvKtPXAQNlek9gTZl+C7AKeBYwGXgYeEdZdi7VII6D23+6TL8MuK1M/0vLMXYDfgbsXPa7FpjUJs9DqJ6k3xl4JnA78MKybA2wZ5ttTqIaD2yHMj8J2JFq5Oq/KLFLB/Mdsu0OwC4tv30V1WClMwZ/R1n2HuDsoeerJa9TyvQ7gQvL9P8GzirTRwC3lOmzgRXATi3rvbFMP30wns/2/ckVSfSq1bZvKdMrqP5Ybs2/2d5g+wGqQvL1El85ZPvL4U/vctmljE/1KuAMVUOwX0f1x316Wf8atx9C4jDgq7YfcfUekK8Af7OVHF8BfNKliajs93nl9/6srLOYqsgNJeBfJN0KfJfqNQl7b+V47QwOaNl6Xg8DPltyuhbYQ9KuZdlS278v09cD75f0PuA5LfHYjqWQRK/a2DK9mSf6+zbxxP/XO46wzeMt84/z5P7CoeMGmeqP9N+7epPcwban2x58u94jw+Q4ljHq1eb4bfcj6cWlCekWSccCb6S62jrE9sFUY6vtyJPPCWx5XoYaPC+t53Wk1zD86ffb/jxwLPB74DuSjtjKsWI7kEIS25s1VE1KAK8d4z5eD38a6PFh2w8D36HqO1BZ9sIa+/k+cJykZ5TRVV9DNXDkSK4G3iFph3KcSVQj9M6QtF9Z583A92zf2FLYlgK7Ur1n4zFJLweeU9a/H9hL0h6lb+WYluNtoGruq/Nb3lhyOhx40Pbvhq5UBgL8ue3zqEaRPajGvqPH5a6t2N58FFgi6c3AtWPcx0OSfgjsArytxP4n1cjRt5ZisoYn/0Hegu2bJV0C3FRCF9r+8VaOfSHwF+U4j1H11/wfSW8FvlgKzI+AT7bZ9jLg65KWA7dQFSBKYfkw1UjHqwfjxSXAJyX9HnjJCHmdDXymNJs9yhPDjA/1eqqbFR4DfgV8eCu/N7YDGf03IiIaSdNWREQ0kkISERGNpJBEREQjKSQREdFICklERDSSQhIREY2kkERERCMpJBER0cj/BwZoVQK7ZfTYAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist( x=[ len(ca) for ca in coauthors.values() ], bins=range(60) )\n",
    "plt.xlabel('number of co-authors')\n",
    "plt.ylabel('number of researchers')\n",
    "plt.xlim(0,51);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have explored this network, let's move to the citation network."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Citations network"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we can look at the citation network. In contrast to the co-authorship network, the citation network is a _directed_ network, where edges can be drawn as arrows. We'll start by expanding the our data about citations into two mappings: \n",
    "\n",
    "* `papers_citing[id]`: papers citing a given paper\n",
    "* `cited_by[id]`: papers cited by a given paper (in other words: its list of references)\n",
    "\n",
    "`papers_citing` will give us the list of a node's incoming links, whereas `cited_by` will give us the list of its outgoing links."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "papers_citing = Citations  # no changes needed, this is what we are storing already in the Citations dataset\n",
    "\n",
    "cited_by = defaultdict(list)\n",
    "\n",
    "for ref, papers_citing_ref in papers_citing.items():\n",
    "    for id in papers_citing_ref:\n",
    "        cited_by[ id ].append( ref )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we are dealing with a subset of the data (the papers mentioning \"fungus\"), `papers_citing` can contain references to papers outside of our subset. On the other hand, the way we created `cited_by`, it will only contain backward references from within our dataset, meaning that it is incomplete with respect to the whole dataset. Nethertheless, we can use this citation network on our subset of fungus-related papers to implement link analysis techniques.\n",
    "\n",
    "Let us now look at an exemplary paper, let's say the one with identifier 29907567. We can now use the `cited_by` mapping to retrieve its (incomplete) list of references:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 references found for paper 29907567\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{29588468: 'Ras hyperactivation versus overexpression: Lessons from Ras dynamics in Candida albicans.',\n",
       " 15302823: 'Ectopic expression of URA3 can influence the virulence phenotypes and proteome of Candida albicans but can be overcome by targeted reintegration of URA3 at the RPS10 locus.',\n",
       " 11722734: 'Ras links cellular morphogenesis to virulence by regulation of the MAP kinase and cAMP signalling pathways in the pathogenic fungus Candida albicans.'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_id = 29907567\n",
    "refs = { id : Summaries[id].title for id in cited_by[paper_id] }\n",
    "print(len(refs), 'references found for paper', paper_id)\n",
    "refs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we lookup the same paper in `papers_citing`, we now see that some of the cited papers are themselves in our dataset, but others are not (shown below as `'??'`):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{32102426: 'The Role of Secretory Pathways in <i>Candida albicans</i> Pathogenesis.',\n",
       " 31440220: '??',\n",
       " 31186458: '??'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{ id : Summaries.get(id,['??'])[0]  for id in papers_citing[paper_id] }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Paper 31186458, for example, is not in our dataset and we do not have any direct information about it, but its repeated occurrence in other papers' citation lists does allow us to reconstruct some of its references. Below is the list of papers in our dataset cited by that paper:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 references identified for the paper with id 31186458\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{29907567: 'Ras signaling activates glycosylphosphatidylinositol (GPI) anchor biosynthesis via the GPI-<i>N</i>-acetylglucosaminyltransferase (GPI-GnT) in <i>Candida albicans</i>.',\n",
       " 15470093: 'Deficiencies in the essential Smp3 mannosyltransferase block glycosylphosphatidylinositol assembly and lead to defects in growth and cell wall biogenesis in Candida albicans.',\n",
       " 15302823: 'Ectopic expression of URA3 can influence the virulence phenotypes and proteome of Candida albicans but can be overcome by targeted reintegration of URA3 at the RPS10 locus.',\n",
       " 12101300: 'GPI7 affects cell-wall protein anchorage in Saccharomyces cerevisiae and Candida albicans.'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_id2 = 31186458\n",
    "refs2 = { id : Summaries[id].title for id in cited_by[paper_id2] }\n",
    "print(len(refs2), 'references identified for the paper with id', paper_id2)\n",
    "refs2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have a better understanding about the data we're dealing with, let us obtain again some basic statistics about our graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of papers in our subset: 54751 (100.00 %)\n",
      "Number of papers cited at least once: 39069 (71.36 %)\n",
      "Number of isolated nodes:  13506 (24.67 %)\n"
     ]
    }
   ],
   "source": [
    "n = len(Ids)\n",
    "print('Number of papers in our subset: {:d} ({:.2f} %)'.format(n, 100.0) )\n",
    "\n",
    "with_citation = [ id for id in Ids if papers_citing[id] != [] ]\n",
    "with_citation_rel = 100. * len(with_citation) / n\n",
    "print('Number of papers cited at least once: {:d} ({:.2f} %)'.format(len(with_citation), with_citation_rel) )\n",
    "\n",
    "isolated = set( id for id in Ids if papers_citing[id] == [] and id not in cited_by )\n",
    "isolated_rel = 100. * len(isolated) / n\n",
    "print('Number of isolated nodes:  {:d} ({:.2f} %)'.format(len(isolated), isolated_rel) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overall number of nodes: 163671 (100.00 %)\n",
      "Number of non-isolated nodes: 150165 (91.75 %)\n",
      "Number of nodes outside our subset: 108920 (66.55 %)\n"
     ]
    }
   ],
   "source": [
    "id_set = set( Ids )\n",
    "citing_set = set( cited_by.keys() )\n",
    "\n",
    "outsiders = citing_set - id_set   # set difference\n",
    "nodes = citing_set | id_set   # set union\n",
    "non_isolated = nodes - isolated   # set difference\n",
    "\n",
    "print('Overall number of nodes: {:d} ({:.2f} %)'.format(len(nodes), 100.0) )\n",
    "\n",
    "non_isolated_rel = 100. * len(non_isolated) / len(nodes)\n",
    "print('Number of non-isolated nodes: {:d} ({:.2f} %)'.format(len(non_isolated), non_isolated_rel) )\n",
    "\n",
    "outsiders_rel = 100. * len(outsiders) / len(nodes)\n",
    "print('Number of nodes outside our subset: {:d} ({:.2f} %)'.format( len(outsiders), outsiders_rel ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overall number of links (citations): 377535 (100.00 %)\n",
      "Citations outside the subset: 278998 (73.90 %)\n"
     ]
    }
   ],
   "source": [
    "all_citations = [ c for citing in papers_citing.values() for c in citing ]\n",
    "outsider_citations = [ c for citing in papers_citing.values() for c in citing if c in outsiders ]\n",
    "\n",
    "print('Overall number of links (citations): {:d} ({:.2f} %)'.format(len(all_citations), 100.0) )\n",
    "\n",
    "outsider_citations_rel = 100. * len(outsider_citations) / len(all_citations)\n",
    "print('Citations outside the subset: {:d} ({:.2f} %)'.format(len(outsider_citations), outsider_citations_rel) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us now find our which 10 papers are the most cited in our dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1002/yea.320101310>New heterologous modules for classical or PCR-based gene disruptions in Saccharomyces cerevisiae.</a></strong><br>1994. Wach A, Brachat A, Pöhlmann R, Philippsen P<br>[ID: 7747518]<br>Citation count: 1065"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>Aspergillus fumigatus and aspergillosis.</strong><br>1999. Latgé JP<br>[ID: 10194462]<br>Citation count: 572"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nature01554>The genome sequence of the filamentous fungus Neurospora crassa.</a></strong><br>2003. Galagan JE, Calvo SE, Borkovich KA, Selker EU, Read ND, Jaffe D, FitzHugh W, Ma LJ, Smirnov S, Purcell S, Rehman B, Elkins T, Engels R, Wang S, Nielsen CB, Butler J, Endrizzi M, Qui D, Ianakiev P, Bell-Pedersen D, ...<br>[ID: 12712197]<br>Citation count: 517"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1111/j.1364-3703.2011.00783.x>The Top 10 fungal pathogens in molecular plant pathology.</a></strong><br>2012. Dean R, Van Kan JA, Pretorius ZA, Hammond-Kosack KE, Di Pietro A, Spanu PD, Rudd JJ, Dickman M, Kahmann R, Ellis J, Foster GD<br>[ID: 22471698]<br>Citation count: 509"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1073/pnas.0601456103>A high-throughput gene knockout procedure for Neurospora reveals functions for multiple transcription factors.</a></strong><br>2006. Colot HV, Park G, Turner GE, Ringelberg C, Crew CM, Litvinkova L, Weiss RL, Borkovich KA, Dunlap JC<br>[ID: 16801547]<br>Citation count: 462"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nature08358>Genome sequence and analysis of the Irish potato famine pathogen Phytophthora infestans.</a></strong><br>2009. Haas BJ, Kamoun S, Zody MC, Jiang RH, Handsaker RE, Cano LM, Grabherr M, Kodira CD, Raffaele S, Torto-Alalibo T, Bozkurt TO, Ah-Fong AM, Alvarado L, Anderson VL, Armstrong MR, Avrova A, Baxter L, Beynon J, Boevink PC, Bollmann SR, ...<br>[ID: 19741609]<br>Citation count: 443"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nature08652>High-performance genetically targetable optical neural silencing by light-driven proton pumps.</a></strong><br>2010. Chow BY, Han X, Dobry AS, Qian X, Chuong AS, Li M, Henninger MA, Belfort GM, Lin Y, Monahan PE, Boyden ES<br>[ID: 20054397]<br>Citation count: 441"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nature04332>Genomic sequence of the pathogenic and allergenic filamentous fungus Aspergillus fumigatus.</a></strong><br>2005. Nierman WC, Pain A, Anderson MJ, Wortman JR, Kim HS, Arroyo J, Berriman M, Abe K, Archer DB, Bermejo C, Bennett J, Bowyer P, Chen D, Collins M, Coulsen R, Davies R, Dyer PS, Farman M, Fedorova N, Fedorova N, ...<br>[ID: 16372009]<br>Citation count: 428"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1038/nature03449>The genome sequence of the rice blast fungus Magnaporthe grisea.</a></strong><br>2005. Dean RA, Talbot NJ, Ebbole DJ, Farman ML, Mitchell TK, Orbach MJ, Thon M, Kulkarni R, Xu JR, Pan H, Read ND, Lee YH, Carbone I, Brown D, Oh YY, Donofrio N, Jeong JS, Soanes DM, Djonovic S, Kolomiets E, ...<br>[ID: 15846337]<br>Citation count: 421"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong><a href=http://dx.doi.org/10.1073/pnas.88.16.7160>A constant rate of spontaneous mutation in DNA-based microbes.</a></strong><br>1991. Drake JW<br>[ID: 1831267]<br>Citation count: 417"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "citation_count_per_paper = [ (id, len(citations)) for (id,citations) in papers_citing.items() ]\n",
    "sorted_by_citation_count = sorted(citation_count_per_paper, key=lambda i:i[1], reverse=True)\n",
    "\n",
    "for (id, c) in sorted_by_citation_count[:10]:\n",
    "    display_summary(id, extra_text = 'Citation count: ' + str(c))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we can start calculating some interesting network metrics, we will first have a closer look at the Python package that we are going to use for that."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Link Analysis for Search Engines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to use the citation network, we need to be able to perform some complex graph algorithms on it. To make our lives easier, we will use [NetworkX](https://pypi.python.org/pypi/networkx), a Python package for dealing with complex networks:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: networkx in /usr/lib/python3/dist-packages (2.4)\r\n"
     ]
    }
   ],
   "source": [
    "! pip install --user networkx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "G = nx.DiGraph(cited_by)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have a NetworkX Directed Graph stored in `G`, where a node represents a paper, and an edge represents a citation. This means we can now apply the [algorithms](http://networkx.github.io/documentation/networkx-1.10/reference/algorithms.html) and [functions](http://networkx.github.io/documentation/networkx-1.10/reference/functions.html) of NetworkX to our graph:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: \n",
      "Type: DiGraph\n",
      "Number of nodes: 150165\n",
      "Number of edges: 377535\n",
      "Average in degree:   2.5141\n",
      "Average out degree:   2.5141\n",
      "Directed graph: True\n",
      "Density of graph: 1.6742591114399504e-05\n"
     ]
    }
   ],
   "source": [
    "print(nx.info(G))\n",
    "print('Directed graph:', nx.is_directed(G))\n",
    "print('Density of graph:', nx.density(G))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As this graph was generated from citations only, we need to add all isolated nodes (nodes that are not cited and do not cite other papers) as well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "G.add_nodes_from(isolated)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now we get slightly different values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: \n",
      "Type: DiGraph\n",
      "Number of nodes: 163671\n",
      "Number of edges: 377535\n",
      "Average in degree:   2.3067\n",
      "Average out degree:   2.3067\n",
      "Directed graph: True\n",
      "Density of graph: 1.4093420224626821e-05\n"
     ]
    }
   ],
   "source": [
    "print(nx.info(G))\n",
    "print('Directed graph:', nx.is_directed(G))\n",
    "print('Density of graph:', nx.density(G))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to use this package for our tasks."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tasks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Your name:** ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 1\n",
    "\n",
    "Plot the in-degree distribution (that is, the distribution of the number of incoming links; see [here](https://en.wikipedia.org/wiki/Degree_distribution) and [here](http://mathinsight.org/degree_distribution) for more detailed explanations) for the citation network. What can you tell about the shape of this distribution, and what does this tell us about the network?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Answer:** [_Write your answer text here_]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 2\n",
    "\n",
    "Using the [Link Analysis](https://networkx.github.io/documentation/networkx-1.10/reference/algorithms.link_analysis.html) algorithms provided by NetworkX, calculate the PageRank score for all nodes in the citation network, and store the result in a variable called `pageranks`. You can also use the `pagerank_scipy` implementation, which tends to be considerably faster than its regular `pagerank` counterpart, for which you have to install the [SciPy package](http://scipy.org/):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: scipy in /usr/lib/python3/dist-packages (1.3.3)\r\n"
     ]
    }
   ],
   "source": [
    "! pip install --user scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To test this, create first a list of all papers that are cited by exactly 10 other papers (using `papers_citing` defined above). Then sort these papers by their PageRank and retrieve the paper IDs with the maximum and minimum PageRank values. Print these minimum and maximum PageRank values together with their paper IDs.\n",
    "\n",
    "To print PageRank values, you might want to use commands like `print('{:.6f}'.format(var))` to use regular decimal notation with a fixed number of decimal places."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use this:\n",
    "#pageranks = nx.pagerank(G)\n",
    "# or this:\n",
    "#pageranks = nx.pagerank_scipy(G)\n",
    "\n",
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 3\n",
    "\n",
    "Why do the two papers above have such different PageRank values? Write code below to investigate and show the cause of this, and then explain the cause of this difference based on the results generated by your code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**Answer:** [_Write your answer text here_]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 4\n",
    "\n",
    "Copy the scoring function `score_ntn_nnn` from Task 4 of assignment 3. Rename it to `score_ntn_nnn_pagerank` and change its code to incorporate a paper's PageRank score in it's final score, in addition to tf-idf. In other words, the new function should accept a `list` of query tokens and a document ID, and should return a single `float` value that is calculated based on both scores (PageRank and tf-idf). Note that a `tf-idf` function is already provided above. Explain your decision on how to combine the two scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**Answer:** [_Write your answer text here_]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 5\n",
    "\n",
    "Copy the query function `query_ntn_nnn` from Task 4 of assignment 3 (also copy `or_merge`). Rename it to `query_ntn_nnn_pagerank` and change the code to use our new scoring function `score_ntn_nnn_pagerank` from task 4 above. Demonstrate these functions with an example query that returns paper 19741609 as the top result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Submission"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Submit the answers to the assignment via Canvas as a modified version of this Notebook file (file with `.ipynb` extension) that includes your code and your answers.\n",
    "\n",
    "Before submitting, restart the kernel and re-run the complete code (**Kernel > Restart & Run All**), and then check whether your assignment code still works as expected.\n",
    "\n",
    "Don't forget to add your name, and remember that the assignments have to be done **individually**, and that code sharing or copying are **strictly forbidden** and will be punished."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
